Compile failure on operator<< of repeat_n( 0, 10 ) | bounded

The following fails to compile under both Clang and GCC:

~~~~~.cpp
#include <range/v3/all.hpp>

#include <iostream>

int main() {
	std::cerr << ( ranges::view::repeat_n( 0, 10 ) | ranges::view::bounded ) << "\n";
}
~~~~~

Apologies if this is erroneous usage. Start of GCC errors:

~~~~~
In file included from i/range/v3/view.hpp:24:0,
                 from i/range/v3/all.hpp:21,
                 from rangev3_repeat_n_bounded.cpp:1:
i/range/v3/view/bounded.hpp: In instantiation of ‘ranges::v3::bounded_view<Rng>::iterator ranges::v3::bounded_view<Rng>::end() const [with int _concept_requires_85 = 42; typename std::enable_if<((_concept_requires_80 == 43) || ranges::v3::concepts::models<ranges::v3::concepts::View, const Rng>()), int>::type <anonymous> = 0; Rng = ranges::v3::repeat_n_view<int>; ranges::v3::bounded_view<Rng>::iterator = ranges::v3::basic_iterator<ranges::v3::repeat_n_view<int>::cursor, ranges::v3::default_end_cursor>]’:
i/range/v3/begin_end.hpp:129:36:   required from ‘static constexpr decltype (rng.end()) ranges::v3::adl_begin_end_detail::end_fn::impl(Rng&&, int) [with Rng = const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >&; typename meta::v1::detail::_if_<meta::v1::list<std::integral_constant<bool, std::is_lvalue_reference< <template-parameter-1-1> >::value>, int>, std::integral_constant<bool, true> >::type <anonymous> = 0; decltype (rng.end()) = ranges::v3::basic_iterator<ranges::v3::repeat_n_view<int>::cursor, ranges::v3::default_end_cursor>]’
i/range/v3/begin_end.hpp:133:32:   required from ‘constexpr ranges::v3::detail::decay_t<decltype (ranges::v3::adl_begin_end_detail::end_fn::impl(static_cast<Rng&&>(rng), 0))> ranges::v3::adl_begin_end_detail::end_fn::operator()(Rng&&) const [with Rng = const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >&; ranges::v3::detail::decay_t<decltype (ranges::v3::adl_begin_end_detail::end_fn::impl(static_cast<Rng&&>(rng), 0))> = ranges::v3::basic_iterator<ranges::v3::repeat_n_view<int>::cursor, ranges::v3::default_end_cursor>]’
i/range/v3/range_concepts.hpp:74:57:   required by substitution of ‘template<class T> decltype (ranges::v3::detail::valid_expr(model_of<ranges::v3::concepts::Sentinel>(ranges::v3::{anonymous}::end(t), ranges::v3::{anonymous}::begin(t)))) ranges::v3::concepts::Range::requires_(T&&) [with T = const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >]’
i/range/v3/utility/concepts.hpp:113:17:   required by substitution of ‘template<class ... Ts, class Concept, class> meta::v1::apply<meta::v1::quote<meta::v1::lazy::strict_and>, typename meta::v1::detail::transform_<meta::v1::list<typename ranges::v3::detail::base_concepts_of<Concept, void>::type, meta::v1::bind_back<meta::v1::quote<ranges::v3::concepts::models>, Ts ...> >, void>::type> ranges::v3::detail::models_(Concept*) [with Ts = {const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >}; Concept = ranges::v3::concepts::Range; <template-parameter-1-3> = void]’
i/range/v3/utility/concepts.hpp:212:69:   required from ‘struct ranges::v3::concepts::models<ranges::v3::concepts::Range, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >’
i/meta/meta.hpp:1186:56:   required by substitution of ‘template<class ... Bools> using strict_and = meta::v1::and_c<Bools::type::value ...> [with Bools = {ranges::v3::concepts::models<ranges::v3::concepts::Range, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}]’
i/meta/meta.hpp:646:21:   required by substitution of ‘template<template<class ...> class C, class ... Ts, class> meta::v1::detail::defer_if_::result<C, Ts ...> meta::v1::detail::defer_if_::try_() [with C = meta::v1::strict_and; Ts = {ranges::v3::concepts::models<ranges::v3::concepts::Range, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}; <template-parameter-1-3> = std::is_same<meta::v1::integer_sequence<bool, true>, meta::v1::integer_sequence<bool, true> >]’
i/meta/meta.hpp:667:66:   required by substitution of ‘template<template<class ...> class C, class ... Ts> using defer_ = decltype (meta::v1::detail::defer_if_{}.try_<template<class ...> class C, Ts ...>()) [with C = meta::v1::strict_and; Ts = {ranges::v3::concepts::models<ranges::v3::concepts::Range, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}]’
i/meta/meta.hpp:712:16:   required from ‘struct meta::v1::defer<meta::v1::strict_and, ranges::v3::concepts::models<ranges::v3::concepts::Range, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > > >’
i/range/v3/utility/concepts.hpp:211:20:   required from ‘struct ranges::v3::concepts::models<ranges::v3::concepts::View, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >’
i/meta/meta.hpp:1186:56:   required by substitution of ‘template<class ... Bools> using strict_and = meta::v1::and_c<Bools::type::value ...> [with Bools = {ranges::v3::concepts::models<ranges::v3::concepts::View, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >, ranges::v3::concepts::models<ranges::v3::concepts::InputRange, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}]’
i/meta/meta.hpp:646:21:   required by substitution of ‘template<template<class ...> class C, class ... Ts, class> meta::v1::detail::defer_if_::result<C, Ts ...> meta::v1::detail::defer_if_::try_() [with C = meta::v1::strict_and; Ts = {ranges::v3::concepts::models<ranges::v3::concepts::View, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >, ranges::v3::concepts::models<ranges::v3::concepts::InputRange, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}; <template-parameter-1-3> = std::is_same<meta::v1::integer_sequence<bool, true, true>, meta::v1::integer_sequence<bool, true, true> >]’
i/meta/meta.hpp:667:66:   required by substitution of ‘template<template<class ...> class C, class ... Ts> using defer_ = decltype (meta::v1::detail::defer_if_{}.try_<template<class ...> class C, Ts ...>()) [with C = meta::v1::strict_and; Ts = {ranges::v3::concepts::models<ranges::v3::concepts::View, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >, ranges::v3::concepts::models<ranges::v3::concepts::InputRange, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >}]’
i/meta/meta.hpp:712:16:   required from ‘struct meta::v1::defer<meta::v1::strict_and, ranges::v3::concepts::models<ranges::v3::concepts::View, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >, ranges::v3::concepts::models<ranges::v3::concepts::InputRange, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > > >’
i/range/v3/utility/concepts.hpp:211:20:   required from ‘struct ranges::v3::concepts::models<ranges::v3::concepts::InputView, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> > >’
i/range/v3/view_interface.hpp:276:39:   required by substitution of ‘template<bool B, class Stream, class D, int _concept_requires_276, typename std::enable_if<((_concept_requires_276 == 43) || ranges::v3::concepts::models<ranges::v3::concepts::InputView, const D>()), int>::type <anonymous> > Stream& ranges::v3::operator<<(Stream&, const ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >&) [with bool B = true; Stream = std::basic_ostream<char>; D = ranges::v3::bounded_view<ranges::v3::repeat_n_view<int> >; int _concept_requires_276 = 42; typename std::enable_if<((_concept_requires_276 == 43) || ranges::v3::concepts::models<ranges::v3::concepts::InputView, const D>()), int>::type <anonymous> = 0]’
rangev3_repeat_n_bounded.cpp:6:73:   required from here
i/range/v3/view/bounded.hpp:88:50: error: no matching function for call to ‘ranges::v3::basic_iterator<ranges::v3::repeat_n_view<int>::cursor, ranges::v3::default_end_cursor>::basic_iterator(<brace-enclosed initializer list>)’
                 return iterator{ranges::end(rng_)};
                                                  ^
~~~~~

Full Clang/GCC errors, with/without colouring :

[gcc_errors.plain.txt](https://github.com/ericniebler/range-v3/files/594082/gcc_errors.plain.txt)
[gcc_errors.colour.txt](https://github.com/ericniebler/range-v3/files/594080/gcc_errors.colour.txt)
[clang_errors.plain.txt](https://github.com/ericniebler/range-v3/files/594083/clang_errors.plain.txt)
[clang_errors.colour.txt](https://github.com/ericniebler/range-v3/files/594081/clang_errors.colour.txt)

Many thanks for your continuing work on this wonderful library.

--- a/include/range/v3/view/bounded.hpp
+++ b/include/range/v3/view/bounded.hpp
@@ -14,7 +14,6 @@
 #define RANGES_V3_VIEW_BOUNDED_HPP
 
 #include <type_traits>
-#include <range/v3/detail/satisfy_boost_range.hpp>
 #include <range/v3/begin_end.hpp>
 #include <range/v3/iterator_range.hpp>
 #include <range/v3/range_concepts.hpp>
@@ -22,6 +21,7 @@
 #include <range/v3/range_traits.hpp>
 #include <range/v3/size.hpp>
 #include <range/v3/view_interface.hpp>
+#include <range/v3/detail/satisfy_boost_range.hpp>
 #include <range/v3/utility/common_iterator.hpp>
 #include <range/v3/utility/functional.hpp>
 #include <range/v3/utility/iterator_concepts.hpp>
@@ -36,62 +36,53 @@ namespace ranges
         /// \addtogroup group-views
         /// @{
 
+        namespace detail
+        {
+            /// \cond
+            template<typename R>
+            using RA_and_Sized =
+                meta::strict_and<RandomAccessRange<R>, SizedRange<R>>;
+
+            template<typename R>
+            using bounded_iterator_t = meta::if_<RA_and_Sized<R>,
+                range_iterator_t<R>,
+                common_iterator<range_iterator_t<R>, range_sentinel_t<R>>>;
+            /// \endcond
+        }
+
         template<typename Rng>
         struct bounded_view
           : view_interface<bounded_view<Rng>, range_cardinality<Rng>::value>
         {
         private:
-            using RA_and_Sized =
-                meta::strict_and<RandomAccessRange<Rng>, SizedRange<Rng>>;
-            using base_iterator_t = range_iterator_t<Rng>;
-            using base_sentinel_t = range_sentinel_t<Rng>;
+            CONCEPT_ASSERT(Range<Rng>());
+            CONCEPT_ASSERT(!BoundedRange<Rng>());
+
             Rng rng_;
 
-            base_sentinel_t end_(std::false_type)
+            range_sentinel_t<Rng> end_(std::false_type)
             {
                 return ranges::end(rng_);
             }
-            base_iterator_t end_(std::true_type)
+            range_iterator_t<Rng> end_(std::true_type)
+            {
+                return ranges::begin(rng_) + ranges::distance(rng_);
+            }
+            template<typename R = Rng const, CONCEPT_REQUIRES_(View<R>())>
+            range_sentinel_t<R> end_(std::false_type) const
+            {
+                return ranges::end(rng_);
+            }
+            template<typename R = Rng const, CONCEPT_REQUIRES_(View<R>())>
+            range_iterator_t<R> end_(std::true_type) const
             {
                 return ranges::begin(rng_) + ranges::distance(rng_);
             }
         public:
-            using iterator = meta::if_<RA_and_Sized,
-                base_iterator_t,
-                common_iterator<base_iterator_t, base_sentinel_t>>;
-
             bounded_view() = default;
             explicit bounded_view(Rng rng)
               : rng_(detail::move(rng))
             {}
-            iterator begin()
-            {
-                return iterator{ranges::begin(rng_)};
-            }
-            iterator end()
-            {
-                return iterator{end_(RA_and_Sized{})};
-            }
-            CONCEPT_REQUIRES(SizedRange<Rng>())
-            range_size_t<Rng> size()
-            {
-                return ranges::size(rng_);
-            }
-            CONCEPT_REQUIRES(View<Rng const>())
-            iterator begin() const
-            {
-                return iterator{ranges::begin(rng_)};
-            }
-            CONCEPT_REQUIRES(View<Rng const>())
-            iterator end() const
-            {
-                return iterator{ranges::end(rng_)};
-            }
-            CONCEPT_REQUIRES(SizedRange<Rng const>())
-            range_size_t<Rng> size() const
-            {
-                return ranges::size(rng_);
-            }
             Rng & base()
             {
                 return rng_;
@@ -99,6 +90,36 @@ namespace ranges
             Rng const & base() const
             {
                 return rng_;
+            }
+
+            detail::bounded_iterator_t<Rng> begin()
+            {
+                return detail::bounded_iterator_t<Rng>{ranges::begin(rng_)};
+            }
+            detail::bounded_iterator_t<Rng> end()
+            {
+                return detail::bounded_iterator_t<Rng>{end_(detail::RA_and_Sized<Rng>{})};
+            }
+            CONCEPT_REQUIRES(SizedRange<Rng>())
+            range_size_t<Rng> size()
+            {
+                return ranges::size(rng_);
+            }
+
+            template<typename R = Rng const, CONCEPT_REQUIRES_(View<R>())>
+            detail::bounded_iterator_t<R> begin() const
+            {
+                return detail::bounded_iterator_t<R>{ranges::begin(rng_)};
+            }
+            template<typename R = Rng const, CONCEPT_REQUIRES_(View<R>())>
+            detail::bounded_iterator_t<R> end() const
+            {
+                return detail::bounded_iterator_t<R>{end_(detail::RA_and_Sized<R>{})};
+            }
+            CONCEPT_REQUIRES(SizedRange<Rng const>())
+            range_size_t<Rng> size() const
+            {
+                return ranges::size(rng_);
             }
         };
 

--- a/test/view/bounded.cpp
+++ b/test/view/bounded.cpp
@@ -14,8 +14,9 @@
 #include <sstream>
 #include <range/v3/core.hpp>
 #include <range/v3/view/bounded.hpp>
-#include <range/v3/view/delimit.hpp>
 #include <range/v3/view/counted.hpp>
+#include <range/v3/view/delimit.hpp>
+#include <range/v3/view/repeat_n.hpp>
 #include "../simple_test.hpp"
 #include "../test_utils.hpp"
 
@@ -23,49 +24,85 @@ int main()
 {
     using namespace ranges;
 
-    std::stringstream sinx("1 2 3 4 5 6 7 8 9 1 2 3 4 5 6 7 8 9 1 2 3 4 42 6 7 8 9 ");
-    auto rng0 = istream<int>(sinx) | view::delimit(42) | view::bounded;
-    ::models<concepts::BoundedView>(rng0);
-    ::models<concepts::InputIterator>(rng0.begin());
-    CONCEPT_ASSERT(Same<typename std::iterator_traits<decltype(rng0.begin())>::iterator_category,
-                        std::input_iterator_tag>());
-    ::models_not<concepts::ForwardIterator>(rng0.begin());
-    ::check_equal(rng0, {1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4});
+    {
+        std::stringstream sinx("1 2 3 4 5 6 7 8 9 1 2 3 4 5 6 7 8 9 1 2 3 4 42 6 7 8 9 ");
+        auto rng1 = istream<int>(sinx) | view::delimit(42); // | view::bounded;
+        ::models_not<concepts::BoundedRange>(rng1);
+        ::models<concepts::InputRange>(rng1);
+        ::models_not<concepts::ForwardRange>(rng1);
+        auto const& crng1 = rng1;
+        ::models_not<concepts::Range>(crng1);
+        auto rng2 = rng1 | view::bounded;
+        ::models<concepts::BoundedView>(rng2);
+        ::models<concepts::InputIterator>(rng2.begin());
+        CONCEPT_ASSERT(Same<typename std::iterator_traits<decltype(rng2.begin())>::iterator_category,
+                            std::input_iterator_tag>());
+        ::models_not<concepts::ForwardIterator>(rng2.begin());
+        ::check_equal(rng2, {1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4});
+    }
 
     std::vector<int> v{1,2,3,4,5,6,7,8,9,0,42,64};
-    auto rng1 = v | view::delimit(42) | view::bounded;
-    ::models<concepts::BoundedView>(rng1);
-    ::models_not<concepts::SizedView>(rng1);
-    ::models<concepts::ForwardIterator>(rng1.begin());
-    ::models_not<concepts::BidirectionalIterator>(rng1.begin());
-    auto const & crng1 = rng1;
-    auto i = rng1.begin(); // non-const
-    auto j = crng1.begin(); // const
-    j = i;
-    ::check_equal(rng1, {1, 2, 3, 4, 5, 6, 7, 8, 9, 0});
+    {
+        auto rng1 = v | view::delimit(42) | view::bounded;
+        ::models<concepts::BoundedView>(rng1);
+        ::models_not<concepts::SizedView>(rng1);
+        ::models<concepts::ForwardIterator>(rng1.begin());
+        ::models_not<concepts::BidirectionalIterator>(rng1.begin());
+        auto const & crng1 = rng1;
+        auto i = rng1.begin(); // non-const
+        auto j = crng1.begin(); // const
+        j = i;
+        ::check_equal(rng1, {1, 2, 3, 4, 5, 6, 7, 8, 9, 0});
+    }
 
-    std::list<int> l{1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0};
-    auto rng3 = view::counted(l.begin(), 10) | view::bounded;
-    ::models<concepts::BoundedView>(rng3);
-    ::models<concepts::SizedView>(rng3);
-    ::models<concepts::ForwardIterator>(rng3.begin());
-    ::models_not<concepts::BidirectionalIterator>(rng3.begin());
-    ::models<concepts::SizedSentinel>(rng3.begin(), rng3.end());
-    auto b = begin(rng3);
-    auto e = end(rng3);
-    CHECK((e-b) == 10);
-    CHECK((b-e) == -10);
-    CHECK((e-e) == 0);
-    CHECK((next(b)-b) == 1);
+    {
+        std::list<int> l{1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0};
+        auto rng3 = view::counted(l.begin(), 10) | view::bounded;
+        ::models<concepts::BoundedView>(rng3);
+        ::models<concepts::SizedView>(rng3);
+        ::models<concepts::ForwardIterator>(rng3.begin());
+        ::models_not<concepts::BidirectionalIterator>(rng3.begin());
+        ::models<concepts::SizedSentinel>(rng3.begin(), rng3.end());
+        auto b = begin(rng3);
+        auto e = end(rng3);
+        CHECK((e-b) == 10);
+        CHECK((b-e) == -10);
+        CHECK((e-e) == 0);
+        CHECK((next(b)-b) == 1);
 
-    // Pass-through of already-bounded ranges is OK:
-    rng3 = rng3 | view::bounded;
+        // Pass-through of already-bounded ranges is OK:
+        rng3 = rng3 | view::bounded;
+    }
 
-    auto rng4 = view::counted(begin(v), 8) | view::bounded;
-    ::models<concepts::BoundedView>(rng4);
-    ::models<concepts::SizedView>(rng4);
-    ::models<concepts::RandomAccessIterator>(begin(rng4));
-    ::check_equal(rng4, {1, 2, 3, 4, 5, 6, 7, 8});
+    {
+        auto rng4 = view::counted(begin(v), 8) | view::bounded;
+        ::models<concepts::BoundedView>(rng4);
+        ::models<concepts::SizedView>(rng4);
+        ::models<concepts::RandomAccessIterator>(begin(rng4));
+        ::check_equal(rng4, {1, 2, 3, 4, 5, 6, 7, 8});
+    }
+
+    {
+        // Regression test for issue#504:
+        auto rng1 = view::repeat_n( 0, 10 );
+        ::models_not<concepts::BoundedView>(rng1);
+        ::models<concepts::RandomAccessView>(rng1);
+        ::models<concepts::SizedView>(rng1);
+        auto const& crng1 = rng1;
+        ::models<concepts::RandomAccessView>(crng1);
+        ::models<concepts::SizedView>(crng1);
+
+        auto rng2 = rng1 | view::bounded;
+        ::models<concepts::BoundedView>(rng2);
+        ::models<concepts::RandomAccessView>(rng2);
+        ::models<concepts::SizedView>(rng2);
+        auto const& crng2 = rng2;
+        ::models<concepts::BoundedView>(crng2);
+        ::models<concepts::RandomAccessView>(crng2);
+        ::models<concepts::SizedView>(crng2);
+
+        ::check_equal(rng2, {0, 0, 0, 0, 0, 0, 0, 0, 0, 0});
+    }
 
     return ::test_result();
 }

